src/parser/paladin/retribution/modules/core/Judgment.js

```diff
@@ -17,38 +21,73 @@ class Judgment extends Analyzer {
   divineStormConsumptions = 0;
   justicarsVengeanceConsumptions = 0;
   judgmentsApplied = 0;
+  judgmentsOverwritten = 0;
+  judgmentTargets = [];
 
   on_byPlayer_applydebuff(event) {
     const spellId = event.ability.guid;
     if (spellId !== SPELLS.JUDGMENT_DEBUFF.id) {
       return;
     }
+    // The event doesn't specify instance on the first target of that type
+    let targetInstance = event.targetInstance;
+    if (targetInstance === undefined) {
+      targetInstance = 1;
+    }
     this.judgmentsApplied += 1;
+    const judgmentTarget = { targetID: event.targetID, targetInstance: targetInstance, timestamp: event.timestamp };
+    this.judgmentTargets.push(judgmentTarget);
   }
 
-  on_byPlayer_damage(event) {
-    const enemy = this.enemies.getEntity(event);
+  on_byPlayer_refreshdebuff(event) {
     const spellId = event.ability.guid;
-
-    if (!enemy) {
+    if (spellId !== SPELLS.JUDGMENT_DEBUFF.id) {
       return;
     }
-    if (!enemy.hasBuff(SPELLS.JUDGMENT_DEBUFF.id, null, 250)) {
+    let targetInstance = event.targetInstance;
+    if (targetInstance === undefined) {
+      targetInstance = 1;
+    }
+    this.judgmentsOverwritten += 1;
+    this.judgmentTargets.forEach((judgmentTarget, index) => {
+      if (judgmentTarget.targetID === event.targetID && judgmentTarget.targetInstance === targetInstance) {
+        judgmentTarget.timestamp = event.timestamp;
+      }
+    });
+  }
+
+  on_byPlayer_damage(event) {
+    const spellId = event.ability.guid;
+    if (!HOLY_POWER_ATTACKS.includes(spellId)) {
       return;
     }
-    switch (spellId) {
-      case SPELLS.TEMPLARS_VERDICT_DAMAGE.id:
-        this.templarsVerdictConsumptions += 1;
-        break;
-      case SPELLS.DIVINE_STORM_DAMAGE.id:
-        this.divineStormConsumptions += 1;
-        break;
-      case SPELLS.JUSTICARS_VENGEANCE_TALENT.id:
-        this.justicarsVengeanceConsumptions += 1;
-        break;
-      default:
-        break;
+    let targetInstance = event.targetInstance;
+    if (targetInstance === undefined) {
+      targetInstance = 1;
     }
+    this.judgmentTargets.forEach((judgmentTarget, index) => {
+      // remove expired judgments
+      const expirationTimestamp = judgmentTarget.timestamp + JUDGMENT_DURATION;
```

Comment:
This should trigger a removebuff, relying on that should be more reliable. I think this could break with lag.

Reply:
removebuff is triggered correctly already. Whether the judgmentTarget object is removed here or in a removebuff event trigger is entirely irrelevant, it just has to be done before it's counted as consumed.

Reply:
It is relevant though as `removebuff` will always match exactly how it happened in-game, while this is just a guess at when that might be.

Reply:
In that case please leave a comment in the code explaining why you didn't use `removebuff`